aten/src/ATen/native/cuda/SegmentReduce.cu

```diff
@@ -372,27 +382,59 @@ Tensor _segment_reduce_cuda_backward_kernel(
   return grad_input;
 }
 
-Tensor _segment_reduce_cuda_kernel(
-    SegmentReductionType reduction,
-    const Tensor& data,
-    const Tensor& lengths,
-    int64_t axis,
-    const c10::optional<Scalar>& initial) {
-  // data and lengths should be contiguous from the call to .contiguous in segment_reduce_kernel
-  TORCH_CHECK(data.is_contiguous(), "Expected data to be contiguous.");
-  TORCH_CHECK(lengths.is_contiguous(), "Expected lengths to be contiguous.");
-  axis = lengths.dim() - 1;
-  int64_t segment_count = lengths.size(axis);
-  int64_t lengths_stride_axis = lengths.stride(axis);
+Tensor _segment_reduce_lengths_backward_cuda_kernel(
+  const Tensor& grad_contig,
+  const Tensor& output_contig,
+  const Tensor& data_contig,
+  SegmentReductionType reduction,
+  const Tensor& lengths_contig,
+  int64_t axis,
+  const c10::optional<Scalar>& initial) {
+  return _segment_reduce_lengths_offsets_backward_cuda_kernel(
+    grad_contig, output_contig, data_contig, reduction, lengths_contig, axis, initial, /*is_offsets_like=*/false);
+}
+
+Tensor _segment_reduce_offsets_backward_cuda_kernel(
+  const Tensor& grad_contig,
+  const Tensor& output_contig,
+  const Tensor& data_contig,
+  SegmentReductionType reduction,
+  const Tensor& offsets_contig,
+  int64_t axis,
+  const c10::optional<Scalar>& initial) {
+  return _segment_reduce_lengths_offsets_backward_cuda_kernel(
+    grad_contig, output_contig, data_contig, reduction, offsets_contig, axis, initial, /*is_offsets_like=*/true);
+}
+
+Tensor _segment_reduce_lengths_offsets_cuda_kernel(
+  SegmentReductionType reduction,
+  const Tensor& data,
+  const Tensor& lengths_or_offsets,
+  int64_t axis,
+  const c10::optional<Scalar>& initial,
+  bool is_offsets_like) {
```

Comment:
Tried to make `is_offsets_like` a template variable like `is_scatter_like` in ScatterGatherKernel, but oddly enough that caused errors like `<data> : undeclared identifier` (and similarly for all the other variables in the nested lambda) specifically on Windows. I've noticed that windows has problems with capturing variables in nested lambdas from [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L860-L861). But in this case removing the template and passing `is_offsets_like` (and changing absolutely nothing else the lambda is still nested) seemed to fix it but I'm not sure why this is the case. Do you have any thoughts on how else I could fix this?